docs/source/specs/typespec/main.tsp

```diff
@@ -0,0 +1,251 @@
+}
+
+interface CreateResource<BodyType> {
+    create<ModelType>(@body body: BodyType): CreatedResponseForModel<ModelType> | 
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse400 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ListResource<T> {
+    list(...ListQueryParams): T[] |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ReadResource<T> {
+    read(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): T |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse404 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface UpdateResource<B, R, E> {
+    update(
+        @doc("Unique identification")
+        @path uuid: UUID;
+        @body body: B
+    ): OkResponseForModel<R> |
+    Errors.ErrorResponse400AlreadyExist<E> |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface DeleteResource {
+    delete(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): NoContentResponseResourceDeleted |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+@route("/workspaces")
+@tag("Workspaces")
+@doc("Operations about workspaces")
+namespace Workspaces {
+    model WorkspaceForRead {
+        @doc("Workspace A")
+        name: string = "Workspace A";
+        @doc("Description of Workspace A")
+        description?: string = "Description of Workspace A";
+    }
+
+    model Workspace {
+        @key uuid: UUID;
+        ...WorkspaceForRead;
+        ...Timestamps;
+    }
+
+    @error
+    model ErrorMessage400WorkspaceAlreadyExist {
+        detail?: string = "Workspace already exist in tenant",
+        status?: string = "400",
+        source?: string = "workspace",
+    }
+
+    alias CreateResourceAlias = CreateResource<WorkspaceForRead>;
+    @doc("Create workspace in tenant")
+    @summary("Create workspace in tenant")
+    @post op create is CreateResourceAlias.create<Workspace>;
+
+    model WorkspaceListResponse is ListResponse<Workspace>;
+    alias ListResourceAlias = ListResource<WorkspaceListResponse>;
+    @doc("By default, responses are sorted in ascending order by workspace name")
+    @summary("List workspaces for tenant")
+    @get op list is ListResourceAlias.list;
+
+    alias ReadResourceAlias = ReadResource<Workspace>;
+    @doc("Get a workspace in tenant")
+    @summary("Get a workspace in tenant")
+    @get op read is ReadResourceAlias.read;
+
+    alias UpdateResourceAlias = UpdateResource<WorkspaceForRead, Workspace, ErrorMessage400WorkspaceAlreadyExist>;
+    @doc("Edit a workspace")
+    @summary("Edit the workspace name or description")
+    @put op update is UpdateResourceAlias.update;
+
+    @doc("Delete a workspace")
+    @summary("Delete the workspace")
+    @delete op delete is DeleteResource.delete;
```

Comment:
For delete we probably want an error/problem type for "cannot delete, it isn't empty."

Reply:
Added 👍